subsecs.h

```diff
@@ -0,0 +1,282 @@
+/*
+ * subsecs.h
+ *
+ *  Created on: 16 Sep 2016
+ *      Author: acolomitchi
+ */
+
+#ifndef SUBSECS_H_
+#define SUBSECS_H_
+
+#include <iostream>
+#include <limits>
+#include <locale>
+#include <string>
+
+#if defined(_MSC_VER) && ! defined(__clang__)
+// MS cl compiler.
+#  define CONSTDATA const
+#  define CONSTCD11
+#  define CONSTCD14
+#  define NOEXCEPT _NOEXCEPT
+#elif __cplusplus >= 201402
+// C++14
+#  define CONSTDATA constexpr
+#  define CONSTCD11 constexpr
+#  define CONSTCD14 constexpr
+#  define NOEXCEPT noexcept
+#else
+// C++11
+#  define CONSTDATA constexpr
+#  define CONSTCD11 constexpr
+#  define CONSTCD14
+#  define NOEXCEPT noexcept
+#endif
+
+
+namespace date {
+namespace detail {
+
+
+// Like std::enable_if, but for values
+template <bool, typename T, T v>
+struct v_enable_if;
+
+template <typename T, T v>
+struct v_enable_if<false, T, v> {
+};
+
+template <typename T, T v>
+struct v_enable_if<true,T,v> {
+  static CONSTDATA T value=v;
+};
+
+template <std::uintmax_t val, std::uintmax_t base>
+struct ceil_log_struct {
+
+  // need one level of indirection to arrest infinite recursion upon
+  // checking '(value<b)' stop condition
+  template <bool, std::uintmax_t v, std::uintmax_t b=10>
+  struct nl_len_tag;
+
+  // recursion will stop when the (val < b), the literal length will be 1
+  template <std::uintmax_t v, std::uintmax_t b>
+  struct nl_len_tag<false, v, b> {
+    static CONSTDATA unsigned value=1;
+  };
+
+  // if val>=b, call numliteral_len with val/b and add one to the returned value
+  template <std::uintmax_t v, std::uintmax_t b>
+  struct nl_len_tag<true, v, b> {
+    static CONSTDATA unsigned value=ceil_log_struct<v/b, b>::value + 1;
+  };
+
+  static CONSTDATA unsigned value=
+      v_enable_if<
+          (base>1), // don't accept base of 0 or 1, they don't make sense
+          unsigned,
+          nl_len_tag<(val>base), val, base>::value
+      >::value
+  ;
+};
+
+// ------
+// ***Warning: note the unusual order of (power, base) for the parameters
+// *** due to the default val for the base
+
+template <unsigned long exponent, std::uintmax_t base=10>
+struct pow_struct
+{
+private:
+  static CONSTDATA std::uintmax_t at_half_pow=pow_struct<exponent / 2, base>::value;
+public:
+  static CONSTDATA std::uintmax_t value=
+      at_half_pow*at_half_pow*(exponent % 2 ? base : 1)
+  ;
+};
+
+template <std::uintmax_t base>
+struct pow_struct<1, base>
+{
+  static CONSTDATA long long value=base;
+};
+
+
+template <std::uintmax_t base>
+struct pow_struct<0,base>
+{
+  static CONSTDATA long long value=1;
+};
+
+#define POW10(width) (pow_struct<(width), 10>::value)
+#define CEIL_LOG10(val) (ceil_log_struct<(val),10>::value)
```

Comment:
Since these don't have to be general purpose utilities, we might want to drop these macros and hard-code the constant `10` into the structs.